### ISSUE TITLE
[Benchmark] Fix redundant ccls in lm_head

### ISSUE THREAD

**Issue by mvasiljevicTT** [tenstorrent/tt-xla#5511]
GLM 4.7 (and probably other models as well) currently end with this pattern:

<img width="600"  alt="Image" src="https://github.com/user-attachments/assets/7ca1f407-cd64-4df9-94ea-9a766b3dcbf6" />

with massive logits tensor making `gather-shard-gather` round-trip over batch dimensions. 

This is happening because logits are annotated as fully-replicated, even though in a perf run outputs are batch-sharded. 

Quick fix here is to [modify the lm_head sharding](https://github.com/tenstorrent/tt-xla/commit/9867378a80da6ef72129060908882fc095ea6ab7#diff-6e5a8960831b592893a9a8714226e0e42cddbcb01315ea8f8882d82925f337b4R464-R467), but that will further tangle already complicated code flow.

We should find a more elegant way to clean this up.

**Comment by mvasiljevicTT** [comment]
### Root cause

The lm_head logits are annotated **fully-replicated**. In a perf run the sampled tokens are batch-sharded, so a replicated constraint forces a redundant **DP `all_gather` + `mesh_partition`** round-trip on the full `[64, 1, 151552]` logits tensor around the `argmax`.

Why the replication was there at all: `argmax` reduces over the **vocab** dim, which is TP-sharded, and tt-mlir miscompiles `argmax` over a sharded dim (local argmax → `all_gather` indices → `sum`). So the vocab dim genuinely has to be gathered — but full replication gathered vocab **and** batch, and only the vocab gather is needed.

### Change (`050c3b831`)

Constrain logits to **batch-sharded + vocab-replicated**, applied directly before the `argmax` in `LLMSamplingWrapper` instead of via an `lm_head` forward hook. This keeps only the required vocab gather, drops the batch round-trip, and co-locates the constraint with the op that needs it (works for any model, no `lm_head` assumption).

### Perf — GLM-4.7, galaxy, bs64, isl128 (1 layer, to isolate the lm_head)

| | decode ms/token | tok/s/user | TTFT ms |
|---|---|---|---|
| before (replicated) | 79.1 | 12.6 | 123.8 |
| after (batch-sharded) | 32.9 | 30.4 | 78.7 |

The round-trip costs **~46 ms/decode-token** (and ~45 ms of TTFT). It is layer-independent (lm_head only), so that ~46 ms is an absolute saving — here a **2.4x** decode speedup with the rest of the model minimized. Also verified on Deepseek-v3.2-exp.

_Measured as e2e per-token decode latency; the sharding constraint is the only difference between the two runs, so the delta is the round-trip._

**Comment by mvasiljevicTT** [comment]
### CI perf benchmarks

Branch: `mvasiljevic/glm-opt1-ci-galaxy` (commit `050c3b831`), built against tt-mlir `5170ccfa` (ArgMax/MoE operand workarounds). `sh-runner=false`.

| Target | Run | Result |
|---|---|---|
| galaxy-wh-6u | [28673788714](https://github.com/tenstorrent/tt-xla/actions/runs/28673788714) | 7/8 — **GLM-4.7 ✅**, llama-70b ✅, etc. |
| qb2-blackhole | [28673791221](https://github.com/tenstorrent/tt-xla/actions/runs/28673791221) | 24/26 |
| galaxy-bh | — | no perf tests registered for this arch |

**Conclusion:** the change is validated — the target `glm_4_7_tp_galaxy_4_layers` and every other galaxy LLM benchmark pass. None of the failures are attributable to the logits-sharding change; all are device hangs / fabric-topology issues on the runners:

- `deepseek_v3_2_exp` (galaxy) — `TT_THROW: device timeout ... device is unrecoverable` (cores 18-27 never finish) after the perf run completed; re-running. Passes PCC locally with this change.
- `gpt_oss_120b` (qb2) — consistent device timeout, fails pre-change too.
- `gpt_oss_20b` (qb2) — device hang.
- `falcon3_7b` (qb2) — `topology_mapper` mapping failure; **passed on rerun** (flake).
